docs/runtime-model.md

```diff
@@ -0,0 +1,356 @@
+Task`][current]. This handle is obtained by reading a thread-local variable set
+by the executor.
+
+Some external event (data received on the network, background thread completing
+a computation, etc...) will result in `my_resource` becoming ready to produce
+its value. At that point, the logic that readies `my_resource` will call
+[`notify`] on the task handle obtained from [`task::current`][current]. This
+signals the readiness change to the executor, which then schedules the task for
+execution.
+
+#### `Async::Ready`
+
+Any function that returns `Async` must adheare to the [contract][contract]. When
+`NotReady` is returned, the current task **must** have been registered for
+notification on readiness change. The implication for resources is discussed in
+the above section. For task logic, this means that `NotReady` cannot be returned
+unless a resource has returned `NotReady`. By doing this, the
+[contract][contract] transitively upheld. The current task is registered for
+notification because `NotReady` has been received from the resource.
+
+Great care must be taken to avoiding returning `NotReady` without having
+received `NotReady` from a resource. For example, the following task
+implementation results in the task never completing.
+
+```rust
+
+enum BadTask {
+    First(Resource1),
+    Second(Resource2),
+}
+
+impl Future for BadTask {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        use self::BadTask::*;
+
+        match *self {
+            First(ref mut resource) => {
+                let value = try_ready!(resource.poll());
+                *self = Second(Resource2::new(value));
+                Ok(Async::NotReady)
+            }
+            Second(ref mut resource) => {
+                try_ready!(resource.poll());
+                Ok(Async::Ready(()))
+            }
+        }
+    }
+}
+```
+
+The problem with the above implementation is that `Ok(Async::NotReady)` is
+returned right after transitioning the state to `Second`. During this
+transition, no resource has returned `NotReady`. When the task itself returns
+`NotReady`, it has violated the [contract][contract] as the task will **not** be
+notified in the future.
+
+This situation is generally resolved by adding a loop:
+
+```rust
+fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    use self::BadTask::*;
+
+    loop {
+        match *self {
+            First(ref mut resource) => {
+                let value = try_ready!(resource.poll());
+                *self = Second(Resource2::new(value));
+            }
+            Second(ref mut resource) => {
+                try_ready!(resource.poll());
+                return Ok(Async::Ready(()));
+            }
+        }
+    }
+}
+```
+
+One way to think about it is that a task's `poll` function **must not**
+returned until it is unable to make any further progress due to its resources
+not being ready.
```

Comment:
The first time I read this, it sounded like it contradicted the earlier point about tasks never blocking. It might be useful to add a small sentence afterwards indicating that if you do need to return without waiting on resources, you need to yield, which is explained in the next section